internal/csi-common/utils.go

```diff
@@ -306,3 +306,46 @@ func IsBlockMultiNode(caps []*csi.VolumeCapability) (bool, bool) {
 
 	return isBlock, isMultiNode
 }
+
+// IsMultiWriterBlockRWOFile validates the volume capability slice against the access modes and access type.
+// if the capability is of multi write the first return value will be set to true and if the request
+// is of type block, the second return value will be set to true. If it is of type RWO and file mode, the third return
+// will be set to true. If the access mode is reader only the last return value has been set to true.
+// Otherwise, false values are returned for all the return arguments.
+func IsMultiWriterBlockRWOFile(caps []*csi.VolumeCapability) (bool, bool, bool, bool) {
+	var multiWriter bool // this flag has been set and returned after validating multi writer caps regardless of
+	// single or multi node access mode. The caps check is agnostic to whether it is a filesystem or block
+	// mode volume.
+	var block bool   // this flag has been set and returned if the passed in capability is of block volume mode.
+	var rwoFile bool // this flag has been set if the volume is of file mode and if the capabilities are of RWO
+	// kind, ie SINGLE NODE but flexible to have one or more writers. This is also used as a validation in caller
+	// to preserve the backward compatibility we had with file mode RWO volumes.
+	var roOnly bool // this flag has been set only when the access mode is `READER ONLY`.
+	for _, cap := range caps {
+		if cap.AccessMode != nil {
+			switch cap.AccessMode.Mode { //nolint:exhaustive // only check what we want
+			case csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY,
+				csi.VolumeCapability_AccessMode_SINGLE_NODE_READER_ONLY:
+				roOnly = true
+			case csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER,
+				csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER:
+				multiWriter = true
+			}
+			// to preserve backward compatibility we allow RWO filemode, ideally SINGLE_NODE_WRITER check is good enough,
+			// however more granular level check could help us in future, so keeping it here as an additional measure.
+			if cap.GetMount() != nil {
+				switch cap.AccessMode.Mode { //nolint:exhaustive // only check what we want
+				case csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+					csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER,
+					csi.VolumeCapability_AccessMode_SINGLE_NODE_SINGLE_WRITER:
```

Comment:
instead of having more values in single case, have multiple cases statements with fallthrough?

Reply:
kept this model as it is a common standard as well. thanks..